RecoTracker/LSTCore/interface/alpaka/LSTESData.h

```diff
@@ -0,0 +1,94 @@
+#define LSTESData_H
+
+#ifdef LST_IS_CMSSW_PACKAGE
+#include "RecoTracker/LSTCore/interface/alpaka/Constants.h"
+#else
+#include "Constants.h"
+#endif
+
+#include "HeterogeneousCore/AlpakaInterface/interface/CopyToDevice.h"
+
+#include <filesystem>
+#include <memory>
+
+namespace SDL {
+
+  struct pixelMap;
+
+  template <typename>
+  class TiltedGeometry;
+
+  template <typename>
+  class ModuleConnectionMap;
+  using MapPLStoLayer = std::array<std::array<ModuleConnectionMap<Dev>, 4>, 3>;
+
+  template <typename>
+  struct modulesBuffer;
+
+  template <typename>
+  class EndcapGeometryHost;
+
+  template <typename>
+  class EndcapGeometry;
+
+  template <typename TDev>
+  struct LSTESHostData;
+
+  // FIXME: This shouldn't be a templated struct
+  template <>
+  struct LSTESHostData<Dev> {
+    std::shared_ptr<const MapPLStoLayer> mapPLStoLayer;
+    std::shared_ptr<const EndcapGeometryHost<Dev>> endcapGeometry;
+    std::shared_ptr<const TiltedGeometry<Dev>> tiltedGeometry;
+    std::shared_ptr<const ModuleConnectionMap<Dev>> moduleConnectionMap;
+
+    LSTESHostData(std::shared_ptr<MapPLStoLayer> mapPLStoLayerIn,
+                  std::shared_ptr<EndcapGeometryHost<Dev>> endcapGeometryIn,
+                  std::shared_ptr<TiltedGeometry<Dev>> tiltedGeometryIn,
+                  std::shared_ptr<ModuleConnectionMap<Dev>> moduleConnectionMapIn)
+        : mapPLStoLayer(mapPLStoLayerIn),
+          endcapGeometry(endcapGeometryIn),
+          tiltedGeometry(tiltedGeometryIn),
+          moduleConnectionMap(moduleConnectionMapIn) {}
+  };
+
+  template <typename TDev>
+  struct LSTESDeviceData {
+    uint16_t nModules;
+    uint16_t nLowerModules;
+    unsigned int nPixels;
+    std::shared_ptr<const modulesBuffer<TDev>> modulesBuffers;
+    std::shared_ptr<const EndcapGeometry<TDev>> endcapGeometry;
+    std::shared_ptr<const pixelMap> pixelMapping;
+
+    LSTESDeviceData(uint16_t nModulesIn,
+                    uint16_t nLowerModulesIn,
+                    unsigned int nPixelsIn,
+                    std::shared_ptr<modulesBuffer<TDev>> modulesBuffersIn,
+                    std::shared_ptr<EndcapGeometry<TDev>> endcapGeometryIn,
+                    std::shared_ptr<pixelMap> pixelMappingIn)
+        : nModules(nModulesIn),
+          nLowerModules(nLowerModulesIn),
+          nPixels(nPixelsIn),
+          modulesBuffers(modulesBuffersIn),
+          endcapGeometry(endcapGeometryIn),
+          pixelMapping(pixelMappingIn) {}
+  };
+
+  std::unique_ptr<LSTESHostData<Dev>> loadAndFillESHost();
+  std::unique_ptr<LSTESDeviceData<Dev>> loadAndFillESDevice(SDL::QueueAcc& queue, const LSTESHostData<Dev>* hostData);
+
+}  // namespace SDL
+
+namespace cms::alpakatools {
+  template <>
+  struct CopyToDevice<SDL::LSTESHostData<SDL::Dev>> {
+    template <typename TQueue>
+    static auto copyAsync(TQueue& queue, SDL::LSTESHostData<SDL::Dev> const& hostData) {
+      return std::make_unique<SDL::LSTESHostData<SDL::Dev>>(hostData);
```

Comment:
This operation would need some explanation (also in the code) on how making a shallow copy of `SDL::LSTESHostData<SDL::Dev>` (i.e. copying only the `shared_ptr`s) constitutes as "copy to device". Does the source `hostData` already hold pointers to device memory? Does the returned object actually hold pointers to host data?

Reply:
@ariostas 
was `CopyToDevice<SDL::LSTESHostData<SDL::Dev>>` even needed in this case? I don't think any of the `LSTESHostData` data is used directly in device kernels; it's all consumed in the CPU code

Reply:
This is needed to pass the ES data that was constructed by `produceHost` into `produceDevice`. This is the specific line that requires it, and I couldn't find a way to get around it even though there is no need to copy any data. The data is still only on the host and is used to initialize the actual device ES data.

https://github.com/SegmentLinking/cmssw/blob/dc4e61f6918237954d7a89c215a0ac877a374807/RecoTracker/LST/plugins/alpaka/LSTModulesDevESProducer.cc#L48

This is also the reason why I ended up needing to add the weird line to `ES_ModulesDev.cc`

Reply:
So if `SDL::LSTESHostData<SDL::Dev>` is not really intended to be copied to the device by the framework, it should be produced in a separate ESProducer that inherits from `edm::ESProducer`. Ideally the ESProducer would be also defined outside of `ALPAKA_ACCELERATOR_NAMESPACE`, that would require addressing this FIXME
https://github.com/cms-sw/cmssw/blob/dc4e61f6918237954d7a89c215a0ac877a374807/RecoTracker/LSTCore/interface/alpaka/LSTESData.h#L38-L40
(well, _technically_ always instantiating it with `alpaka::DevCpu` would work too)

On the other hand, I see `SDL::loadAndFillESHost()` (called from `LSTModulesDevESProducer::produceHost()`) loads some information from files, and `loadAndFillESDevice()` (called from `LSTModulesDevESProducer::produceDevice()`) loads some more information from files. What motivates this separation?

For data copies the recommended way is to create a host-side data structure in `LSTModulesDevESProducer::produceHost()`, specialize `cms::alpakatools::CopyToDevice` for how that data structure is copied to device, and let the framework to handle the actual copy. I.e. the `produceHost()` would produce an object that holds all the data loaded from the files by `SDL::loadAndFillESHost()` and `loadAndFillESDevice()`.